### PR TITLE
fix(approval reminder): exclude solved tickets

### PR DESCRIPTION
**Pull request by Rom1-B**
Notifications were sent on all tickets with no status filtering (resolved or closed)

Feat: https://github.com/glpi-project/glpi/pull/11623

| Q             | A
| ------------- | ---
| Bug fix?      | yes/no
| New feature?  | yes/no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !28109

### CODE REVIEW

**Comment by trasher** [comment]
Maybe this need a test

File: tests/functional/CommonITILValidationCron.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        $this->login();
+
+        $CFG_GLPI['use_notifications']  = true;
+
+        // update entity
+        $entity = new \Entity();
+        $entity->update([
+            'id' => 0,
+            'approval_reminder_repeat_interval' => 1,
+        ]);
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:55]:
```suggestion
        $this->boolean(
            $entity->update([
                'id' => 0,
                'approval_reminder_repeat_interval' => 1,
            ])
        )->isTrue();
```

File: tests/functional/CommonITILValidationCron.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+        // create ticket
+        $ticket = new \Ticket();
+        $ticket_id = $ticket->add([
+            'name' => 'Ticket',
+            'content' => 'Ticket',
+        ]);
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:62]:
```suggestion
        ]);
        $this->integer($ticket_id)->isGreaterThan(0);
```

File: tests/functional/CommonITILValidationCron.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        // create ticket validation
+        $ticket_validation = new \TicketValidation();
+        $ticket_validation_id = $ticket_validation->add([
+            'tickets_id'      => $ticket_id,
+            'itemtype_target' => 'User',
+            'items_id_target' => getItemByTypeName('User', TU_USER, true),
+        ]);
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:70]:
```suggestion
        ]);
        $this->integer($ticket_validation_id)->isGreaterThan(0);
```

File: tests/functional/CommonITILValidationCron.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+        // backdate ticket validation
+        $DB->updateOrDie(
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:73]: I'm not sure this is a good idea to use that in tests

File: tests/functional/CommonITILValidationCron.php [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        // verify last reminder date is empty
+        $ticket_validation->getFromDB($ticket_validation_id);
+        $this->string((string)$ticket_validation->fields['last_reminder_date'])->isEmpty();
+
+        // Solve ticket
+        $ticket->update([
+            'id' => $ticket_id,
+            'status' => \Ticket::SOLVED,
+        ]);
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:118]:
```suggestion
        $this->boolean(
            $ticket->update([
                'id' => $ticket_id,
                'status' => \Ticket::SOLVED,
            ])
        )->isTrue();
```

File: tests/functional/CommonITILValidationCron.php [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+        // run cron
+        \CommonITILValidationCron::cronApprovalReminder($crontask);
+
+        // verify last reminder date is set
+        $ticket_validation->getFromDB($ticket_validation_id);
+        $this->string($ticket_validation->fields['last_reminder_date'])->isNotEmpty();
+
+        // reset last reminder date
+        $DB->updateOrDie(
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:100]: I'm not sure this is a good idea to use that in tests

File: tests/functional/CommonITILValidationCron.php [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        $ticket->update([
+            'id' => $ticket_id,
+            'status' => \Ticket::SOLVED,
+        ]);
+
+        // run cron
+        \CommonITILValidationCron::cronApprovalReminder($crontask);
+
+        // verify last reminder date is empty
+        $ticket_validation->getFromDB($ticket_validation_id);
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:124]:
```suggestion
        $this->boolean($ticket_validation->getFromDB($ticket_validation_id))->isTrue();
```

File: tests/functional/CommonITILValidationCron.php [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        $this->string((string)$ticket_validation->fields['last_reminder_date'])->isEmpty();
+
+        // Close ticket
+        $ticket->update([
+            'id' => $ticket_id,
+            'status' => \Ticket::CLOSED,
+        ]);
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:131]:
```suggestion
        $this->boolean(
            $ticket->update([
                'id' => $ticket_id,
                'status' => \Ticket::CLOSED,
            ])
        )->isTrue();
```

File: tests/functional/CommonITILValidationCron.php [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+        // run cron
+        \CommonITILValidationCron::cronApprovalReminder($crontask);
+
+        // verify last reminder date is empty
+        $ticket_validation->getFromDB($ticket_validation_id);
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:137]:
```suggestion
        $this->boolean($ticket_validation->getFromDB($ticket_validation_id))->isTrue();
```

File: tests/functional/CommonITILValidationCron.php [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        // Close ticket
+        $ticket->update([
+            'id' => $ticket_id,
+            'status' => \Ticket::CLOSED,
+        ]);
+
+        // run cron
+        \CommonITILValidationCron::cronApprovalReminder($crontask);
+
+        // verify last reminder date is empty
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:136]: Test does not check date. Either test or comment is wrong

File: tests/functional/CommonITILValidationCron.php [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        $this->string((string)$ticket_validation->fields['last_reminder_date'])->isEmpty();
+
+        // Close ticket
+        $ticket->update([
+            'id' => $ticket_id,
+            'status' => \Ticket::CLOSED,
+        ]);
+
+        // run cron
+        \CommonITILValidationCron::cronApprovalReminder($crontask);
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:134]: You should test returned cron status.

**Review by trasher — CHANGES_REQUESTED** [review]

File: tests/functional/CommonITILValidationCron.php [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        // backdate ticket validation
+        $DB->update(
+            \TicketValidation::getTable(),
+            [
+                'submission_date' => date('Y-m-d H:i:s', strtotime('-1 day')),
+            ],
+            [
+                'id' => $ticket_validation_id,
+            ]
+        );
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:85]:
```suggestion
        $this->boolean(
            $DB->update(
                \TicketValidation::getTable(),
                [
                    'submission_date' => date('Y-m-d H:i:s', strtotime('-1 day')),
                ],
                [
                    'id' => $ticket_validation_id,
                ]
            )
        )->isTrue();
```

File: tests/functional/CommonITILValidationCron.php [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        // reset last reminder date
+        $DB->update(
+            \TicketValidation::getTable(),
+            [
+                'last_reminder_date' => null,
+            ],
+            [
+                'id' => $ticket_validation_id,
+            ]
+        );
```
Comment (trasher) [inline tests/functional/CommonITILValidationCron.php:113]:
```suggestion
        $this->boolean(
            $DB->update(
                \TicketValidation::getTable(),
                [
                    'last_reminder_date' => null,
                ],
                [
                    'id' => $ticket_validation_id,
                ]
            )
        )->isTrue();
```

**Review by trasher — APPROVED** [review]

**Comment by cedric-anne** [comment]
Please rebase to ensure tests are made on a most recent version of main branch.

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**